Animated menu

Torpedo99 opened the pull request:
# Description
Adds an animated menu to the game.
# Features
### Asteroids
Random asteroids (varying in size, texture and color) are spawned randomly from any side of the screen and move at a random velocity, colliding with each other. At any time, a fixed number of asteroids (currently 10), can be found on the screen.
### Ships
Ships can be seen flying from the right top corner of the screen to the bottom left. The ship is chosen at random from a module containing a [menuBackgroundShipConfig.json](https://github.com/Torpedo99/DestinationSol/blob/animatedMenu/modules/core/assets/configs/menuBackgroundShipConfig.json)

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Torpedo99 commented:
Added a way to spawn ships in the background, you would just need to add the name of the ship you want to spawn (in order of display) to the list.
The game will then iterate through the list and make the ship move from top right to bottom middle, despawn it and spawn the next ship moving in a similar pattern.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

NicholasBatesNZ commented on engine/src/main/java/org/destinationsol/game/drawables/DrawableManager.java:
```diff
@@ -23,6 +23,7 @@
 import com.badlogic.gdx.utils.Array;
 import com.badlogic.gdx.utils.OrderedMap;
 import org.destinationsol.common.DebugCol;
+import org.destinationsol.common.SolColor;
```
What's this doing here? :thinking:

Torpedo99 replied:
Slipped in by mistake :P

NicholasBatesNZ commented on engine/src/main/java/org/destinationsol/menu/background/BackgroundShipManager.java:
```diff
@@ -0,0 +1,100 @@
+                add("core:minerBoss");
+                add("core:desertSmall");
+                add("core:desertOrbiter");
+            }
```
This would be much cooler if it instead read from a config file. That way a module can add its own ships if it so desires.

Torpedo99 replied:
Modules can now add their ships to the menu by using a `menuBackgroundShipConfig.json`

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Torpedo99 commented:
"Zoomed out" camera for the sprite batch, sprites are now drawn at a much normal size. 
Also redid some of the randomizing code.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

NicholasBatesNZ commented on engine/src/main/java/org/destinationsol/menu/MainMenuScreen.java:
```diff
@@ -26,6 +26,7 @@
 import org.destinationsol.assets.audio.OggMusicManager;
 import org.destinationsol.common.SolColor;
 import org.destinationsol.game.DebugOptions;
+import org.destinationsol.menu.background.MenuBackgroundManager;
```
What's this doing here?

Torpedo99 replied:
Residue that I forgot to clean after shifting stuff to SolApplication, removed it now.

NicholasBatesNZ commented on engine/src/main/java/org/destinationsol/menu/background/MenuBackgroundAsteroidManager.java:
```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 MovingBlocks
```
`2020`. Also while we're here, this class needs some javadocs please

NicholasBatesNZ commented on engine/src/main/java/org/destinationsol/menu/background/MenuBackgroundManager.java:
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2019 MovingBlocks
```
`2020` and javadocs

NicholasBatesNZ commented on engine/src/main/java/org/destinationsol/menu/background/MenuBackgroundObject.java:
```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2019 MovingBlocks
```
`2020` and javadocs

NicholasBatesNZ commented on engine/src/main/java/org/destinationsol/menu/background/MenuBackgroundShipManager.java:
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2019 MovingBlocks
```
`2020` and javadocs

NicholasBatesNZ commented on modules/core/assets/configs/menuBackgroundShipConfig.json:
```diff
@@ -0,0 +1,10 @@
+    "minerSmall",
+    "minerBoss",
+    "desertSmall",
+    "desertOrbiter"
```
Please ensure the module is added to these. A module should be allowed to add a ship from another module. You'd also have to handle potential duplicates :+1:

Torpedo99 replied:
Oh, I was wondering if I should do that, but when you said 
> That way a module can add its own ships if it so desires.

I thought modules would only be able to add their own ships :P, I'll change it now

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Torpedo99 commented on engine/src/main/java/org/destinationsol/SolApplication.java:
```diff
@@ -196,7 +201,9 @@ private void update() {
         inputManager.update(this);
 
-        if (solGame != null) {
+        if (solGame == null) {
```
Yes, that maybe a problem in the future, anyway, I found out it was better to keep the animation calls where the background texture is drawn, which includes most of the menu screen classes

Torpedo99 commented on engine/src/main/java/org/destinationsol/SolApplication.java:
```diff
@@ -111,6 +113,7 @@ public void create() {
         musicManager = new OggMusicManager(options);
         soundManager = new OggSoundManager(context);
         inputManager = new SolInputManager(soundManager, context);
+        ;
```
I thought intellij autoformat would fix that kind of stuff, oh well.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Torpedo99 commented:
Background animation update and draws calls are now handled individually depending on whether a screen requires it or not. For now, all Menu Screens (except LoadingScreen.java, since that would just freeze the animation while it loads up SolGame) draw the animated menu.

Modules can now refer to ships of other modules in their `menuBackgroundShipConfig.json` config file.
Also added requested javadocs and updated copyright dates.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!
